File: src/commands/delete.ts
import { Command } from 'commander';
import { getTurbopufferClient } from '../client.js';
import chalk from 'chalk';
import * as readline from 'readline';
import { debugLog } from '../utils/debug.js';

/**
 * Prompts the user for input
 */
function prompt(question: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

export function createDeleteCommand(): Command {
  const deleteCmd = new Command('delete')
    .alias('rm')
    .description('Delete namespace(s)')
    .option('-n, --namespace <name>', 'Namespace to delete')
    .option('--all', 'Delete all namespaces')
    .option('--prefix <prefix>', 'Delete all namespaces starting with prefix')
    .option('-r, --region <region>', 'Override the region (e.g., aws-us-east-1, gcp-us-central1)')
    .action(async (options: { namespace?: string; all?: boolean; prefix?: string; region?: string }) => {
      const client = getTurbopufferClient(options.region);

      try {
        // Validate that exactly one option is provided
        const optionsCount = [options.namespace, options.all, options.prefix].filter(Boolean).length;

        if (optionsCount === 0) {
          console.error(chalk.red('Error: You must specify either -n <namespace>, --all, or --prefix <prefix>'));
          console.log('\nUsage:');
          console.log('  tpuff delete -n <namespace>       Delete a specific namespace');
          console.log('  tpuff delete --all                Delete all namespaces');
          console.log('  tpuff delete --prefix <prefix>    Delete all namespaces starting with prefix');
          process.exit(1);
        }

        if (optionsCount > 1) {
          console.error(chalk.red('Error: Cannot use multiple deletion options together'));
          console.log(chalk.gray('Please use only one of: -n, --all, or --prefix'));
          process.exit(1);
        }

        if (options.namespace) {
          // Delete single namespace
          const namespace = options.namespace;

          console.log(chalk.yellow(`\n⚠️  You are about to delete namespace: ${chalk.bold(namespace)}`));
          console.log(chalk.gray('This action cannot be undone.\n'));

          const answer = await prompt('Are you sure? (y/n): ');

          if (answer.toLowerCase() !== 'y' && answer.toLowerCase() !== 'yes') {
            console.log(chalk.gray('Deletion cancelled.'));
            process.exit(0);
          }

          // Delete the namespace
          console.log(chalk.gray(`\nDeleting namespace ${namespace}...`));
          const ns = client.namespace(namespace);

          debugLog('Delete Parameters', { namespace });
          await ns.deleteAll({ namespace });
          debugLog('Delete Response', 'Success');

          console.log(chalk.green(`✓ Namespace ${chalk.bold(namespace)} deleted successfully!`));

        } else if (options.all) {
          // Delete all namespaces
          console.log(chalk.yellow.bold('\n🚨 DANGER ZONE 🚨'));
          console.log(chalk.red('You are about to delete ALL namespaces!'));
          console.log(chalk.gray('This will permanently destroy all your data.\n'));

          // First, list all namespaces
          const page = await client.namespaces();
          debugLog('Namespaces API Response', page);
          const namespaces = page.namespaces;

          if (namespaces.length === 0) {
            console.log(chalk.gray('No namespaces found. Nothing to delete.'));
            process.exit(0);
          }

          console.log(chalk.yellow(`Found ${namespaces.length} namespace(s):`));
          namespaces.forEach(ns => {
            console.log(chalk.gray(`  - ${ns.id}`));
          });

          console.log(chalk.yellow.bold('\n💀 This is your last chance to back out! 💀'));
          console.log(chalk.gray(`To confirm, please type: ${chalk.bold.red('yolo')}\n`));

          const answer = await prompt('> ');

          if (answer !== 'yolo') {
            console.log(chalk.green('\n✨ Wise choice! Your data lives to see another day.'));
            console.log(chalk.gray('(Phew, that was close!)'));
            process.exit(0);
          }

          // User typed yolo, proceed with deletion
          console.log(chalk.red.bold('\n🎢 YOLO MODE ACTIVATED! 🎢'));
          console.log(chalk.gray('Deleting all namespaces...\n'));

          let successCount = 0;
          let failCount = 0;

          for (const ns of namespaces) {
            try {
              debugLog(`Deleting namespace`, { namespace: ns.id });
              await client.namespace(ns.id).deleteAll({ namespace: ns.id });
              debugLog('Delete Response', 'Success');
              console.log(chalk.gray(`  ✓ Deleted: ${ns.id}`));
              successCount++;
            } catch (error) {
              debugLog('Delete Error', error);
              console.log(chalk.red(`  ✗ Failed to delete: ${ns.id}`));
              console.error(chalk.gray(`    Error: ${error instanceof Error ? error.message : String(error)}`));
              failCount++;
            }
          }

          console.log(chalk.green.bold(`\n🎉 Deletion complete!`));
          console.log(chalk.gray(`Successfully deleted: ${successCount}`));
          if (failCount > 0) {
            console.log(chalk.red(`Failed: ${failCount}`));
          }
          console.log(chalk.gray('\n(Hope you had backups! 😅)'));

        } else if (options.prefix) {
          // Delete namespaces by prefix
          const prefix = options.prefix;

          console.log(chalk.yellow(`\n🔍 Searching for namespaces with prefix: ${chalk.bold(prefix)}`));
          console.log(chalk.gray('(Using case-insensitive matching)\n'));

          // Fetch all namespaces
          const page = await client.namespaces();
          debugLog('Namespaces API Response', page);
          const allNamespaces = page.namespaces;

          // Filter by prefix (case-insensitive)
          const matchingNamespaces = allNamespaces.filter(ns =>
            ns.id.toLowerCase().startsWith(prefix.toLowerCase())
          );

          if (matchingNamespaces.length === 0) {
            console.log(chalk.gray(`No namespaces found with prefix "${prefix}".`));
            console.log(chalk.gray('Nothing to delete.'));
            process.exit(0);
          }

          console.log(chalk.yellow(`Found ${matchingNamespaces.length} namespace(s) matching prefix "${prefix}":`));
          matchingNamespaces.forEach(ns => {
            console.log(chalk.gray(`  - ${ns.id}`));
          });

          console.log(chalk.yellow.bold('\n⚠️  WARNING: This will permanently delete these namespaces!'));
          console.log(chalk.gray(`To confirm, please type the prefix: ${chalk.bold.red(prefix)}\n`));

          const answer = await prompt('> ');

          if (answer.toLowerCase() !== prefix.toLowerCase()) {
            console.log(chalk.green('\n✨ Deletion cancelled.'));
            console.log(chalk.gray('Your data is safe!'));
            process.exit(0);
          }

          // User confirmed, proceed with deletion
          console.log(chalk.red.bold('\n🗑️  Starting deletion...'));
          console.log(chalk.gray(''));

          let successCount = 0;
          let failCount = 0;

          for (const ns of matchingNamespaces) {
            try {
              debugLog(`Deleting namespace`, { namespace: ns.id });
              await client.namespace(ns.id).deleteAll({ namespace: ns.id });
              debugLog('Delete Response', 'Success');
              console.log(chalk.gray(`  ✓ Deleted: ${ns.id}`));
              successCount++;
            } catch (error) {
              debugLog('Delete Error', error);
              console.log(chalk.red(`  ✗ Failed to delete: ${ns.id}`));
              console.error(chalk.gray(`    Error: ${error instanceof Error ? error.message : String(error)}`));
              failCount++;
            }
          }

          console.log(chalk.green.bold(`\n✓ Deletion complete!`));
          console.log(chalk.gray(`Successfully deleted: ${successCount}`));
          if (failCount > 0) {
            console.log(chalk.red(`Failed: ${failCount}`));
          }
        }

      } catch (error) {
        console.error(chalk.red('Error:'), error instanceof Error ? error.message : String(error));
        process.exit(1);
      }
    });

  return deleteCmd;
}

File: src/tpuff/commands/delete.py
"""Delete command for tpuff CLI."""

import sys

import click
from rich.console import Console

from tpuff.client import get_turbopuffer_client
from tpuff.utils.debug import debug_log


console = Console()


def prompt_user(message: str) -> str:
    """Prompt the user for input."""
    return click.prompt(message, default="", show_default=False).strip()


@click.command("delete")
@click.option("-n", "--namespace", help="Namespace to delete")
@click.option("--all", "delete_all", is_flag=True, help="Delete all namespaces")
@click.option("--prefix", help="Delete all namespaces starting with prefix")
@click.option("-r", "--region", help="Override the region (e.g., aws-us-east-1, gcp-us-central1)")
@click.pass_context
def delete(
    ctx: click.Context,
    namespace: str | None,
    delete_all: bool,
    prefix: str | None,
    region: str | None,
) -> None:
    """Delete namespace(s)."""
    client = get_turbopuffer_client(region)

    try:
        # Validate that exactly one option is provided
        options_count = sum([bool(namespace), delete_all, bool(prefix)])

        if options_count == 0:
            console.print("[red]Error: You must specify either -n <namespace>, --all, or --prefix <prefix>[/red]")
            console.print("\nUsage:")
            console.print("  tpuff delete -n <namespace>       Delete a specific namespace")
            console.print("  tpuff delete --all                Delete all namespaces")
            console.print("  tpuff delete --prefix <prefix>    Delete all namespaces starting with prefix")
            sys.exit(1)

        if options_count > 1:
            console.print("[red]Error: Cannot use multiple deletion options together[/red]")
            console.print("[dim]Please use only one of: -n, --all, or --prefix[/dim]")
            sys.exit(1)

        if namespace:
            # Delete single namespace
            console.print(f"\n[yellow]⚠️  You are about to delete namespace: [bold]{namespace}[/bold][/yellow]")
            console.print("[dim]This action cannot be undone.[/dim]\n")

            answer = prompt_user("Are you sure? (y/n)")

            if answer.lower() not in ("y", "yes"):
                console.print("[dim]Deletion cancelled.[/dim]")
                sys.exit(0)

            # Delete the namespace
            console.print(f"\n[dim]Deleting namespace {namespace}...[/dim]")
            ns = client.namespace(namespace)

            debug_log("Delete Parameters", {"namespace": namespace})
            ns.delete_all()
            debug_log("Delete Response", "Success")

            console.print(f"[green]✓ Namespace [bold]{namespace}[/bold] deleted successfully![/green]")

        elif delete_all:
            # Delete all namespaces
            console.print("[bold yellow]\n🚨 DANGER ZONE 🚨[/bold yellow]")
            console.print("[red]You are about to delete ALL namespaces![/red]")
            console.print("[dim]This will permanently destroy all your data.[/dim]\n")

            # First, list all namespaces
            namespaces_iter = client.namespaces()
            debug_log("Namespaces API Response", namespaces_iter)
            namespaces = list(namespaces_iter)

            if not namespaces:
                console.print("[dim]No namespaces found. Nothing to delete.[/dim]")
                sys.exit(0)

            console.print(f"[yellow]Found {len(namespaces)} namespace(s):[/yellow]")
            for ns in namespaces:
                console.print(f"[dim]  - {ns.id}[/dim]")

            console.print("\n[bold yellow]💀 This is your last chance to back out! 💀[/bold yellow]")
            console.print(f"[dim]To confirm, please type: [bold red]yolo[/bold red][/dim]\n")

            answer = prompt_user(">")

            if answer != "yolo":
                console.print("\n[green]✨ Wise choice! Your data lives to see another day.[/green]")
                console.print("[dim](Phew, that was close!)[/dim]")
                sys.exit(0)

            # User typed yolo, proceed with deletion
            console.print("\n[bold red]🎢 YOLO MODE ACTIVATED! 🎢[/bold red]")
            console.print("[dim]Deleting all namespaces...[/dim]\n")

            success_count = 0
            fail_count = 0

            for ns in namespaces:
                try:
                    debug_log("Deleting namespace", {"namespace": ns.id})
                    client.namespace(ns.id).delete_all()
                    debug_log("Delete Response", "Success")
                    console.print(f"[dim]  ✓ Deleted: {ns.id}[/dim]")
                    success_count += 1
                except Exception as e:
                    debug_log("Delete Error", str(e))
                    console.print(f"[red]  ✗ Failed to delete: {ns.id}[/red]")
                    console.print(f"[dim]    Error: {e}[/dim]")
                    fail_count += 1

            console.print("\n[bold green]🎉 Deletion complete![/bold green]")
            console.print(f"[dim]Successfully deleted: {success_count}[/dim]")
            if fail_count > 0:
                console.print(f"[red]Failed: {fail_count}[/red]")
            console.print("[dim]\n(Hope you had backups! 😅)[/dim]")

        elif prefix:
            # Delete namespaces by prefix
            console.print(f"\n[yellow]🔍 Searching for namespaces with prefix: [bold]{prefix}[/bold][/yellow]")
            console.print("[dim](Using case-insensitive matching)[/dim]\n")

            # Fetch all namespaces
            namespaces_iter = client.namespaces()
            debug_log("Namespaces API Response", namespaces_iter)
            all_namespaces = list(namespaces_iter)

            # Filter by prefix (case-insensitive)
            matching_namespaces = [
                ns for ns in all_namespaces if ns.id.lower().startswith(prefix.lower())
            ]

            if not matching_namespaces:
                console.print(f'[dim]No namespaces found with prefix "{prefix}".[/dim]')
                console.print("[dim]Nothing to delete.[/dim]")
                sys.exit(0)

            console.print(f'[yellow]Found {len(matching_namespaces)} namespace(s) matching prefix "{prefix}":[/yellow]')
            for ns in matching_namespaces:
                console.print(f"[dim]  - {ns.id}[/dim]")

            console.print("\n[bold yellow]⚠️  WARNING: This will permanently delete these namespaces![/bold yellow]")
            console.print(f"[dim]To confirm, please type the prefix: [bold red]{prefix}[/bold red][/dim]\n")

            answer = prompt_user(">")

            if answer.lower() != prefix.lower():
                console.print("\n[green]✨ Deletion cancelled.[/green]")
                console.print("[dim]Your data is safe![/dim]")
                sys.exit(0)

            # User confirmed, proceed with deletion
            console.print("\n[bold red]🗑️  Starting deletion...[/bold red]")
            console.print("")

            success_count = 0
            fail_count = 0

            for ns in matching_namespaces:
                try:
                    debug_log("Deleting namespace", {"namespace": ns.id})
                    client.namespace(ns.id).delete_all()
                    debug_log("Delete Response", "Success")
                    console.print(f"[dim]  ✓ Deleted: {ns.id}[/dim]")
                    success_count += 1
                except Exception as e:
                    debug_log("Delete Error", str(e))
                    console.print(f"[red]  ✗ Failed to delete: {ns.id}[/red]")
                    console.print(f"[dim]    Error: {e}[/dim]")
                    fail_count += 1

            console.print("\n[bold green]✓ Deletion complete![/bold green]")
            console.print(f"[dim]Successfully deleted: {success_count}[/dim]")
            if fail_count > 0:
                console.print(f"[red]Failed: {fail_count}[/red]")

    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
        sys.exit(1)
